bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/discovery/BusGatewayUpnpDiscovery.java

```diff
@@ -0,0 +1,229 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.openwebnet.internal.discovery;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.config.discovery.upnp.UpnpDiscoveryParticipant;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.jupnp.model.meta.DeviceDetails;
+import org.jupnp.model.meta.ManufacturerDetails;
+import org.jupnp.model.meta.ModelDetails;
+import org.jupnp.model.meta.RemoteDevice;
+import org.jupnp.model.meta.RemoteDeviceIdentity;
+import org.jupnp.model.types.UDN;
+import org.openhab.binding.openwebnet.OpenWebNetBindingConstants;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link BusGatewayUpnpDiscovery} is responsible for discovering supported BTicino BUS
+ * gateways devices using UPnP. It implements {@link UpnpDiscoveryParticipant}.
+ *
+ * @author Massimo Valla - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = UpnpDiscoveryParticipant.class)
+public class BusGatewayUpnpDiscovery implements UpnpDiscoveryParticipant {
+
+    private final Logger logger = LoggerFactory.getLogger(BusGatewayUpnpDiscovery.class);
+
+    public enum BusGatewayId {
+        MH201("IPscenarioModule", "MH201"),
+        MH202("scheduler", "MH202"),
+        F454("webserver", "F454"),
+        MY_HOME_SERVER1("myhomeserver1", "MYHOMESERVER1"),
+        TOUCH_SCREEN_10("ts10", "TOUCHSCREEN10"),
+        MH200N("lightingcontrolunit", "MH200N");
+
+        private final String value, thingId;
+
+        private BusGatewayId(String value, String thingId) {
+            this.value = value;
+            this.thingId = thingId;
+        }
+
+        public static @Nullable BusGatewayId fromValue(String s) {
+            Optional<BusGatewayId> m = Arrays.stream(values()).filter(val -> s.equals(val.value)).findFirst();
+            if (m == null) {
```

Comment:
`m` cannot be null at this point. The null check could be removed.

Reply:
Changed to m.isPresent(), as m is an Optional